python/ray/data/tests/test_streaming_executor.py

```diff
@@ -333,6 +346,57 @@ def _test_hook(fn, args, strategy):
     assert tasks == ["DEFAULT", "DEFAULT", "SPREAD", "SPREAD"]
 
 
+def test_scheduling_progress_when_output_blocked():
+    # Processing stages should fully finish even if output is completely stalled.
+
+    @ray.remote
+    class Counter:
+        def __init__(self):
+            self.i = 0
+
+        def inc(self):
+            self.i += 1
+
+        def get(self):
+            return self.i
+
+    counter = Counter.remote()
+
+    def func(x):
+        ray.get(counter.inc.remote())
+        return x
+
+    DatasetContext.get_current().use_streaming_executor = True
+    DatasetContext.get_current().execution_options.preserve_order = True
+
+    # Only take the first item from the iterator.
+    it = iter(
+        ray.data.range(100, parallelism=100)
+        .map_batches(func, batch_size=None)
+        .iter_batches(batch_size=None)
+    )
+    next(it)
+    # The pipeline should fully execute even when the output iterator is blocked.
```

Comment:
This is only if we remain under the resource limit, right?

These tests look good for liveness; can you also add a test (if it doesn't already exist) for output backpressure?